metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_classes.py

```diff
@@ -0,0 +1,200 @@
+    reports: List["Report"]
+    datasets: Dict[str, "PowerBIDataset"]
+    report_endorsements: Dict[str, List[str]]
+    dashboard_endorsements: Dict[str, List[str]]
+    scan_result: dict
+
+    def get_urn_part(self):
+        return self.name
+
+    def get_workspace_key(self, platform_name: str) -> PlatformKey:
+        return WorkspaceKey(
+            workspace=self.get_urn_part(),
+            platform=platform_name,
+        )
+
+
+@dataclass
+class DataSource:
+    """
+    PowerBi
+    """
+
+    id: str
+    type: str
+    raw_connection_detail: Dict
+
+    def __members(self):
+        return (self.id,)
+
+    def __eq__(self, instance):
+        return (
+            isinstance(instance, DataSource)
+            and self.__members() == instance.__members()
+        )
+
+    def __hash__(self):
+        return hash(self.__members())
+
+
+@dataclass
+class Table:
+    name: str
+    full_name: str
+    expression: Optional[str]
+
+
+# dataclasses for PowerBi Dashboard
+@dataclass
+class PowerBIDataset:
+    id: str
+    name: Optional[str]
+    webUrl: Optional[str]
+    workspace_id: str
+    # Table in datasets
+    tables: List["Table"]
+    tags: List[str]
+
+    def get_urn_part(self):
+        return f"datasets.{self.id}"
+
+    def __members(self):
+        return (self.id,)
+
+    def __eq__(self, instance):
+        return (
+            isinstance(instance, PowerBIDataset)
+            and self.__members() == instance.__members()
+        )
+
+    def __hash__(self):
+        return hash(self.__members())
+
+
+@dataclass
+class Page:
+    id: str
+    displayName: str
+    name: str
+    order: int
+
+    def get_urn_part(self):
+        return f"pages.{self.id}"
+
+
+@dataclass
+class User:
+    id: str
+    displayName: str
+    emailAddress: str
+    graphId: str
+    principalType: str
+
+    def get_urn_part(self):
+        return f"users.{self.id}"
+
+    def __members(self):
+        return (self.id,)
+
+    def __eq__(self, instance):
+        return isinstance(instance, User) and self.__members() == instance.__members()
+
+    def __hash__(self):
+        return hash(self.__members())
+
+
+@dataclass
+class Report:
+    id: str
+    name: str
+    webUrl: str
+    embedUrl: str
+    description: str
+    dataset: Optional["PowerBIDataset"]
+    pages: List["Page"]
+    users: List["User"]
+    tags: List[str]
+
+    def get_urn_part(self):
+        return f"reports.{self.id}"
+
+
+@dataclass
+class Tile:
+    class CreatedFrom(Enum):
+        REPORT = "Report"
+        DATASET = "Dataset"
+        VISUALIZATION = "Visualization"
+        UNKNOWN = "UNKNOWN"
+
+    id: str
+    title: str
+    embedUrl: str
+    dataset: Optional["PowerBIDataset"]
+    dataset_id: Optional[str]
+    report: Optional[Any]
+    createdFrom: CreatedFrom
+
+    def get_urn_part(self):
+        return f"charts.{self.id}"
+
+
+@dataclass
+class Dashboard:
+    id: str
+    displayName: str
+    embedUrl: str
+    webUrl: str
```

Comment:
Should this be optional?

Reply:
done. Keeping it as is or making it optional doesn't impact the code.